### PR TITLE
Server side filtering for read forward and backwards

### LINKED ISSUE #718: Read all events without including internal ($) streams
I've checked the documentation for EventStore and I have not been able to find a way to read ALL events from a database that does not include events from internal ($) streams. This is particularly important when reading through a network as there is a lot of data wasted.

I think that a flag in the ReadAllEventsForwardAsync (and Backwards) methods could very easily achieve this.

### PR DESCRIPTION
**Pull request by mat-mcloughlin**
From the back of this pull request https://github.com/EventStore/EventStore/pull/1821

Original PR https://github.com/EventStore/EventStore/pull/1972

The original pull request was getting too big so decided to split it down in to reads and subscriptions. This is the read forward and backwards. 

Both TCP/Client and HTTP have been implemented

Fixes #718

### CODE REVIEW

File: src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs [hunk 1]
```diff
@@ -36,18 +53,34 @@ public class AllReader : IAllReader {
 		}
 
 		public IndexReadAllResult ReadAllEventsForward(TFPos pos, int maxCount) {
+			var result = ReadAllEventsForwardInternal(pos, maxCount, maxCount, EventFilter.Get(null));
```
[the remark this answers is not in the document]
Comment (mat-mcloughlin) [reply src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs]: I've updated to add a static property in

File: src/EventStore.Core/Messages/ClientMessage.cs [hunk 2]
```diff
@@ -886,14 +890,167 @@ public class ReadAllEventsBackwardCompleted : ReadResponseMessage {
 			}
 		}
 
-		//Persistent subscriptions
-		public class ConnectToPersistentSubscription : ReadRequestMessage {
+		public class ReadAllEventsForwardFiltered : ReadRequestMessage {
```
[the remark this answers is not in the document]
Comment (mat-mcloughlin) [reply src/EventStore.Core/Messages/ClientMessage.cs]: Added the messages to the controller

File: src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs [hunk 3]
```diff
@@ -258,5 +318,24 @@ public class AllReader : IAllReader {
 			       || (rec.RecordType == LogRecordType.Prepare &&
 			           ((PrepareLogRecord)rec).Flags.HasAnyOf(PrepareFlags.IsCommitted));
 		}
+
+		private struct ReadAllEventsRawResult {
```
Comment (pgermishuys) [inline src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs]: We don't need `ReadAllEventsRawResult` as it's just there to carry the `IsEndOfStream` property to the caller. We can also get rid of the `IndexReadAllFilteredResult` and add the `IsEndOfStream` property to the `IndexReadAllResult`.

Comment (mat-mcloughlin) [reply src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs]: I agree, good idea

**Comment by samholder** [comment]
I don't want to derail this, but I'd like to ask opinions on also having server side filtering based on metadata properties, in addition to the event types.

The use cases I'm thinking of are things like filtering events by tenant or having a way to have events handled by different deployed services (so deploy an instance of an updated service for canary testing and be able to 'route' particular events to that service only via a header added to the event).

Would this be something of interest? (Maybe there are other ways of solving these problems). And would it be something that could be built on the changes made here?

**Comment by jageall** [comment]
@samholder Interesting idea. The problem with doing either metadata or data properties is that they need to be deserialized. There are no restrictions on the format of metadata or data.  This would need some significant design around it. For that reason it won't get added to this PR, and likely won't be in the timeframe for ES6.  If you would like to raise an issue for it, we can discuss it there in more details.

**Pull request merged**